File: python/slide_window/substring_with_words.py
def find_substring(txt: str, words: list) -> list:
    if len(txt) == 0 or len(words) == 0 or len(txt) < len(words) * len(words[0]):
        return []

    word_length = len(words[0])
    result = []
    freqMap = {}

    # Build frequency map for words
    for word in words:
        if word in freqMap:
            freqMap[word] += 1
        else:
            freqMap[word] = 1

    # Outer loop for word-length alignments
    for index in range(word_length):
        start = index
        count = 0
        seen = {}

        # Sliding window
        for end in range(index, len(txt) - word_length + 1, word_length):
            word = txt[end:end + word_length]

            if word in freqMap:
                if word in seen:
                    seen[word] += 1
                else:
                    seen[word] = 1

                count += 1

                # Shrink window if frequency exceeds allowed
                while seen[word] > freqMap[word]:
                    leftword = txt[start:start + word_length]
                    seen[leftword] -= 1
                    count -= 1
                    start += word_length

                # Check if valid substring
                if count == len(words):
                    result.append(start)

            else:
                # Reset window on invalid word
                seen = {}
                count = 0
                start = end + word_length

    return result

File: js/src/WindowSlide/SubstringWithWords.js
export const findSubstring = (s, words) => {
    if (s.length === 0 || words.length === 0 || s.length < words[0].length * words.length) {
        return [];
    }

    let wordLength = words[0].length;
    let mapFreq = {};
    let result = [];

    // Build frequency map for words
    for (let i = 0; i < words.length; i++) {
        if (mapFreq[words[i]] !== undefined) {
            mapFreq[words[i]]++;
        } else {
            mapFreq[words[i]] = 1;
        }
    }

    // Outer loop for word-length alignments
    for (let i = 0; i < wordLength; i++) {
        let start = i;
        let count = 0;
        let seen = {};

        // Inner sliding window
        for (let j = i; j <= s.length - wordLength; j += wordLength) {
            let word = s.substring(j, j + wordLength);

            if (mapFreq[word] !== undefined) {
                // Add word to seen
                if (seen[word] !== undefined) {
                    seen[word]++;
                } else {
                    seen[word] = 1;
                }

                count++;

                // Shrink window if word exceeds frequency
                while (seen[word] > mapFreq[word]) {
                    let leftWord = s.substring(start, start + wordLength);
                    seen[leftWord]--;
                    count--;
                    start += wordLength;
                }

                // Check if valid substring
                if (count === words.length) {
                    result.push(start);
                }
            } else {
                // Reset on invalid word
                start = j + wordLength;
                count = 0;
                seen = {};
            }
        }
    }

    return result;
};
